common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -321,369 +294,340 @@ public boolean awaitUninterruptibly(long timeoutMillis) {
-                return false;
-            }
+        return buf;
+    }
 
-            result = new CauseHolder(cause);
-            if (hasWaiters()) {
-                notifyAll();
-            }
+    protected EventExecutor executor() {
+        return executor;
+    }
+
+    protected void checkDeadLock() {
+        EventExecutor e = executor();
+        if (e != null && e.inEventLoop()) {
+            throw new BlockingOperationException(toString());
         }
-        return true;
     }
 
-    private boolean setSuccess0(V result) {
-        if (isDone()) {
-            return false;
+    /**
+     * Notify a listener that a future has completed.
+     * <p>
+     * This method has a fixed depth of {@link #MAX_LISTENER_STACK_DEPTH} that will limit recursion to prevent
+     * {@link StackOverflowError} and will stop notifying listeners added after this threshold is exceeded.
+     * @param eventExecutor the executor to use to notify the listener {@code l}.
+     * @param future the future that is complete.
+     * @param l the listener to notify.
+     */
+    protected static void notifyListener(
+            EventExecutor eventExecutor, final Future<?> future, final GenericFutureListener<?> l) {
+        if (eventExecutor.inEventLoop()) {
+            final InternalThreadLocalMap threadLocals = InternalThreadLocalMap.get();
+            final int stackDepth = threadLocals.futureListenerStackDepth();
+            if (stackDepth < MAX_LISTENER_STACK_DEPTH) {
+                threadLocals.setFutureListenerStackDepth(stackDepth + 1);
+                try {
+                    notifyListener0(future, l);
+                } finally {
+                    threadLocals.setFutureListenerStackDepth(stackDepth);
+                }
+                return;
+            }
         }
 
-        synchronized (this) {
-            // Allow only once.
-            if (isDone()) {
-                return false;
+        safeExecute(eventExecutor, new OneTimeTask() {
+            @Override
+            public void run() {
+                notifyListener0(future, l);
             }
-            if (result == null) {
-                this.result = SUCCESS;
-            } else {
-                this.result = result;
+        });
+    }
+
+    private void notifyListeners() {
+        // Modifications to listeners should be done in a synchronized block before this, and should be visible here.
+        if (listeners == null) {
+            return;
+        }
+        EventExecutor executor = executor();
+        if (executor.inEventLoop()) {
+            notifyListeners0();
+            return;
+        }
+        safeExecute(executor, new OneTimeTask() {
+            @Override
+            public void run() {
+                notifyListeners0();
             }
-            if (hasWaiters()) {
-                notifyAll();
+        });
+    }
+
+    private void notifyListeners0() {
+        Object listeners;
+        while (!notifyingListeners) {
```

Comment:
Are we concerned about concurrent listener execution or listener reordering? This code allows both.

Consider this series of events:
1. Promise completed; notifyListeners0 scheduled on executor1
2. Executor changes to executor2
3. notifyListeners0 starts running on executor1 and sets `this.listeners = null` and leaves synchronized block. Starts notifying listeners
4. New Listener is added; this.listeners no longer null; notifyListeners0 scheduled on executor2 (or, inEventLoop returns `true`)
5. notifyListeners0 on executor2 notifies listener from step 4
6. notifyListeners0 on executor1 finishes notifying listeners